src/pages/api/sanity/lessons/create.ts

```diff
@@ -13,54 +14,119 @@ type LessonData = {
   }
 }
 
+type SanityReference<SanityType> = {
+  _type: SanityType
+  _ref: string
+}
+
+type SanityReferenceArray<SanityType> = Array<
+  {
+    _key: string
+  } & SanityReference<'reference'>
+>
+
 type SanityVideoResource = {
-  _type: string
+  _type: 'videoResource'
   _id: string
   filename: string
   originalVideoUrl: string
 }
 
 type SanityLesson = {
-  _type: string
+  _type: 'lesson'
+  _id: string
   title: string
   resource: {
     _type: string
     _ref: string
   }
 }
 
+type SanityCourse = {
+  _type: 'course'
+  title: string
+  collaborators: SanityReferenceArray<'collaborator'>
+  lessons: SanityReferenceArray<'lesson'>
+}
+
 const sanityClient = client({
   projectId: 'sb1i5dlc',
   dataset: 'production',
   useCdn: true,
   token: process.env.SANITY_EDITOR_TOKEN,
 })
 
-async function formatSanityMutationForLessons(lessons: LessonData[]) {
+const sanityIdForDocumentType = async (
+  documentType: string,
+): Promise<string> => {
+  const id = await nanoid()
+  return `${documentType}-${id}`
+}
+
+async function formatSanityMutationForLessons(
+  course: CourseData,
+  lessons: LessonData[],
+): Promise<{
+  sanityCourse: SanityCourse
+  sanityLessons: SanityLesson[]
+  sanityResources: SanityVideoResource[]
+}> {
   let sanityLessons: SanityLesson[] = []
   let sanityResources: SanityVideoResource[] = []
 
-  lessons.forEach(async (lesson) => {
-    const videoId = await nanoid()
+  const {title, collaboratorId} = course
 
-    sanityResources.push({
-      _type: 'videoResource',
-      _id: videoId,
-      filename: lesson.fileMetadata.fileName,
-      originalVideoUrl: lesson.fileMetadata.signedUrl,
-    })
+  let sanityCourse: SanityCourse = {
+    _type: 'course',
+    title,
+    lessons: [],
+    collaborators: [],
+  }
+
+  const collaboratorKey = await nanoid()
 
-    sanityLessons.push({
-      _type: 'lesson',
-      title: lesson.title,
-      resource: {
+  if (typeof collaboratorId === 'string') {
+    sanityCourse.collaborators = [
+      {
+        _key: collaboratorKey,
         _type: 'reference',
-        _ref: videoId,
+        _ref: collaboratorId,
       },
-    })
-  })
+    ]
+  }
 
-  return {sanityLessons, sanityResources}
+  await Promise.all(
+    lessons.map(async (lesson) => {
+      const videoId = await sanityIdForDocumentType('videoResource')
+
+      sanityResources.push({
```

Comment:
I had to wrap this whole thing in an `await Promise.all(lessons.map(async () => { ... }))`. Because it is asynchronously adding each lesson to the `sanityLessons` array. Before it was a `forEach` which didn't work because the main thread was moving on with an empty `sanityLessons` array before the inline async functions could process.